 argh.h | 141 ++++++++++++++++++++++++-----------------------------------------
 1 file changed, 51 insertions(+), 90 deletions(-)

diff --git a/argh.h b/argh.h
index 68ce717..42f5404 100644
--- a/argh.h
+++ b/argh.h
@@ -9,6 +9,13 @@
 #include <map>
 #include <cassert>
 
+#include <Core/Container.hpp>
+template<typename T> using arghVector = std::vector<T, Foundation::Core::StlDefaultAllocator<T>>;
+template<typename T> using arghSet = std::set<T, std::less<T>, Foundation::Core::StlDefaultAllocator<T>>;
+template<typename T> using arghMultiset = std::multiset<T, std::less<T>, Foundation::Core::StlDefaultAllocator<T>>;
+template<typename K, typename V> using arghMap = std::map<K, V, std::less<K>, Foundation::Core::StlDefaultAllocator<std::pair<const K, V>>>;
+using arghString = Foundation::Core::String;
+
 namespace argh
 {
    // Terminology:
@@ -18,56 +25,10 @@ namespace argh
    //    2.1: Flags: boolean options =>  (exist ? true : false)
    //    2.2: Parameters: a name followed by a non-option value
 
-#if !defined(__GNUC__) || (__GNUC__ >= 5)
-   using string_stream = std::istringstream;
-#else
-    // Until GCC 5, istringstream did not have a move constructor.
-    // stringstream_proxy is used instead, as a workaround.
-   class stringstream_proxy
-   {
-   public:
-      stringstream_proxy() = default;
-
-      // Construct with a value.
-      stringstream_proxy(std::string const& value) :
-         stream_(value)
-      {}
-
-      // Copy constructor.
-      stringstream_proxy(const stringstream_proxy& other) :
-         stream_(other.stream_.str())
-      {
-         stream_.setstate(other.stream_.rdstate());
-      }
-
-      void setstate(std::ios_base::iostate state) { stream_.setstate(state); }
-
-      // Stream out the value of the parameter.
-      // If the conversion was not possible, the stream will enter the fail state,
-      // and operator bool will return false.
-      template<typename T>
-      stringstream_proxy& operator >> (T& thing)
-      {
-         stream_ >> thing;
-         return *this;
-      }
-
-
-      // Get the string value.
-      std::string str() const { return stream_.str(); }
-
-      std::stringbuf* rdbuf() const { return stream_.rdbuf(); }
-
-      // Check the state of the stream.
-      // False when the most recent stream operation failed
-      explicit operator bool() const { return !!stream_; }
-
-      ~stringstream_proxy() = default;
-   private:
-      std::istringstream stream_;
-   };
-   using string_stream = stringstream_proxy;
-#endif
+   using string_stream =
+       std::basic_istringstream<char, std::char_traits<char>, Foundation::Core::StlDefaultAllocator<char>>;
+   using ostring_stream =
+       std::basic_ostringstream<char, std::char_traits<char>, Foundation::Core::StlDefaultAllocator<char>>;
 
    class parser
    {
@@ -89,32 +50,32 @@ namespace argh
       parser(int argc, const char* const argv[], int mode = PREFER_FLAG_FOR_UNREG_OPTION)
       {  parse(argc, argv, mode); }
 
-      void add_param(std::string const& name);
+      void add_param(arghString const& name);
       void add_params(std::initializer_list<char const* const> init_list);
 
       void parse(const char* const argv[], int mode = PREFER_FLAG_FOR_UNREG_OPTION);
       void parse(int argc, const char* const argv[], int mode = PREFER_FLAG_FOR_UNREG_OPTION);
 
-      std::multiset<std::string>          const& flags()    const { return flags_;    }
-      std::map<std::string, std::string>  const& params()   const { return params_;   }
-      std::vector<std::string>            const& pos_args() const { return pos_args_; }
+      arghMultiset<arghString>          const& flags()    const { return flags_;    }
+      arghMap<arghString, arghString>  const& params()   const { return params_;   }
+      arghVector<arghString>            const& pos_args() const { return pos_args_; }
 
       // begin() and end() for using range-for over positional args.
-      std::vector<std::string>::const_iterator begin() const { return pos_args_.cbegin(); }
-      std::vector<std::string>::const_iterator end()   const { return pos_args_.cend();   }
+      arghVector<arghString>::const_iterator begin() const { return pos_args_.cbegin(); }
+      arghVector<arghString>::const_iterator end()   const { return pos_args_.cend();   }
       size_t size()                                    const { return pos_args_.size();   }
 
       //////////////////////////////////////////////////////////////////////////
       // Accessors
 
       // flag (boolean) accessors: return true if the flag appeared, otherwise false.
-      bool operator[](std::string const& name) const;
+      bool operator[](arghString const& name) const;
 
       // multiple flag (boolean) accessors: return true if at least one of the flag appeared, otherwise false.
       bool operator[](std::initializer_list<char const* const> init_list) const;
 
       // returns positional arg string by order. Like argv[] but without the options
-      std::string const& operator[](size_t ind) const;
+      arghString const& operator[](size_t ind) const;
 
       // returns a std::istream that can be used to convert a positional arg to a typed value.
       string_stream operator()(size_t ind) const;
@@ -125,7 +86,7 @@ namespace argh
 
       // parameter accessors, give a name get an std::istream that can be used to convert to a typed value.
       // call .str() on result to get as string
-      string_stream operator()(std::string const& name) const;
+      string_stream operator()(arghString const& name) const;
 
       // accessor for a parameter with multiple names, give a list of names, get an std::istream that can be used to convert to a typed value.
       // call .str() on result to get as string
@@ -136,7 +97,7 @@ namespace argh
       // Non-string def_val types must have an operator<<() (output stream operator)
       // If T only has an input stream operator, pass the string version of the type as in "3" instead of 3.
       template<typename T>
-      string_stream operator()(std::string const& name, T&& def_val) const;
+      string_stream operator()(arghString const& name, T&& def_val) const;
 
       // same as above but for a list of names. returns the first value to be found.
       template<typename T>
@@ -144,19 +105,19 @@ namespace argh
 
    private:
       string_stream bad_stream() const;
-      std::string trim_leading_dashes(std::string const& name) const;
-      bool is_number(std::string const& arg) const;
-      bool is_option(std::string const& arg) const;
-      bool got_flag(std::string const& name) const;
-      bool is_param(std::string const& name) const;
+      arghString trim_leading_dashes(arghString const& name) const;
+      bool is_number(arghString const& arg) const;
+      bool is_option(arghString const& arg) const;
+      bool got_flag(arghString const& name) const;
+      bool is_param(arghString const& name) const;
 
    private:
-      std::vector<std::string> args_;
-      std::map<std::string, std::string> params_;
-      std::vector<std::string> pos_args_;
-      std::multiset<std::string> flags_;
-      std::set<std::string> registeredParams_;
-      std::string empty_;
+      arghVector<arghString> args_;
+      arghMap<arghString, arghString> params_;
+      arghVector<arghString> pos_args_;
+      arghMultiset<arghString> flags_;
+      arghSet<arghString> registeredParams_;
+      arghString empty_;
    };
 
 
@@ -196,7 +157,7 @@ namespace argh
          if (!(mode & NO_SPLIT_ON_EQUALSIGN))
          {
             auto equalPos = name.find('=');
-            if (equalPos != std::string::npos)
+            if (equalPos != arghString::npos)
             {
                params_.insert({ name.substr(0, equalPos), name.substr(equalPos + 1) });
                continue;
@@ -208,9 +169,9 @@ namespace argh
             argh::parser::SINGLE_DASH_IS_MULTIFLAG & mode && // multi-flag mode
             !is_param(name))                                  // unregistered
          {
-            std::string keep_param;
+            arghString keep_param;
 
-            if (!name.empty() && is_param(std::string(1ul, name.back()))) // last char is param
+            if (!name.empty() && is_param(arghString(1ul, name.back()))) // last char is param
             {
                keep_param += name.back();
                name.resize(name.size() - 1);
@@ -218,7 +179,7 @@ namespace argh
 
             for (auto const& c : name)
             {
-               flags_.emplace(std::string{ c });
+               flags_.emplace(arghString{ c });
             }
 
             if (!keep_param.empty())
@@ -276,10 +237,10 @@ namespace argh
 
    //////////////////////////////////////////////////////////////////////////
 
-   inline bool parser::is_number(std::string const& arg) const
+   inline bool parser::is_number(arghString const& arg) const
    {
       // inefficient but simple way to determine if a string is a number (which can start with a '-')
-      std::istringstream istr(arg);
+      string_stream istr(arg);
       double number;
       istr >> number;
       return !(istr.fail() || istr.bad());
@@ -287,7 +248,7 @@ namespace argh
 
    //////////////////////////////////////////////////////////////////////////
 
-   inline bool parser::is_option(std::string const& arg) const
+   inline bool parser::is_option(arghString const& arg) const
    {
       assert(0 != arg.size());
       if (is_number(arg))
@@ -297,29 +258,29 @@ namespace argh
 
    //////////////////////////////////////////////////////////////////////////
 
-   inline std::string parser::trim_leading_dashes(std::string const& name) const
+   inline arghString parser::trim_leading_dashes(arghString const& name) const
    {
       auto pos = name.find_first_not_of('-');
-      return std::string::npos != pos ? name.substr(pos) : name;
+      return arghString::npos != pos ? name.substr(pos) : name;
    }
 
    //////////////////////////////////////////////////////////////////////////
 
-   inline bool argh::parser::got_flag(std::string const& name) const
+   inline bool argh::parser::got_flag(arghString const& name) const
    {
       return flags_.end() != flags_.find(trim_leading_dashes(name));
    }
 
    //////////////////////////////////////////////////////////////////////////
 
-   inline bool argh::parser::is_param(std::string const& name) const
+   inline bool argh::parser::is_param(arghString const& name) const
    {
       return registeredParams_.count(name);
    }
 
    //////////////////////////////////////////////////////////////////////////
 
-   inline bool parser::operator[](std::string const& name) const
+   inline bool parser::operator[](arghString const& name) const
    {
       return got_flag(name);
    }
@@ -333,7 +294,7 @@ namespace argh
 
    //////////////////////////////////////////////////////////////////////////
 
-   inline std::string const& parser::operator[](size_t ind) const
+   inline arghString const& parser::operator[](size_t ind) const
    {
       if (ind < pos_args_.size())
          return pos_args_[ind];
@@ -342,7 +303,7 @@ namespace argh
 
    //////////////////////////////////////////////////////////////////////////
 
-   inline string_stream parser::operator()(std::string const& name) const
+   inline string_stream parser::operator()(arghString const& name) const
    {
       auto optIt = params_.find(trim_leading_dashes(name));
       if (params_.end() != optIt)
@@ -366,13 +327,13 @@ namespace argh
    //////////////////////////////////////////////////////////////////////////
 
    template<typename T>
-   string_stream parser::operator()(std::string const& name, T&& def_val) const
+   string_stream parser::operator()(arghString const& name, T&& def_val) const
    {
       auto optIt = params_.find(trim_leading_dashes(name));
       if (params_.end() != optIt)
          return string_stream(optIt->second);
 
-      std::ostringstream ostr;
+      ostring_stream ostr;
       ostr.precision(std::numeric_limits<long double>::max_digits10);
       ostr << def_val;
       return string_stream(ostr.str()); // use default
@@ -390,7 +351,7 @@ namespace argh
          if (params_.end() != optIt)
             return string_stream(optIt->second);
       }
-      std::ostringstream ostr;
+      ostring_stream ostr;
       ostr.precision(std::numeric_limits<long double>::max_digits10);
       ostr << def_val;
       return string_stream(ostr.str()); // use default
@@ -413,7 +374,7 @@ namespace argh
    {
       if (pos_args_.size() <= ind)
       {
-         std::ostringstream ostr;
+         ostring_stream ostr;
          ostr.precision(std::numeric_limits<long double>::max_digits10);
          ostr << def_val;
          return string_stream(ostr.str());
@@ -424,7 +385,7 @@ namespace argh
 
    //////////////////////////////////////////////////////////////////////////
 
-   inline void parser::add_param(std::string const& name)
+   inline void parser::add_param(arghString const& name)
    {
       registeredParams_.insert(trim_leading_dashes(name));
    }
