 demo/clusterlod.h | 74 ++++++++++++++++++++++++++++---------------------------
 1 file changed, 38 insertions(+), 36 deletions(-)

diff --git a/demo/clusterlod.h b/demo/clusterlod.h
index 554daaf7..62fdc800 100644
--- a/demo/clusterlod.h
+++ b/demo/clusterlod.h
@@ -11,6 +11,8 @@
 #pragma once
 
 #include <stddef.h>
+#include <Core/Container.hpp>
+template<typename T> using clodVector = std::vector<T, Foundation::Core::StlDefaultAllocator<T>>;
 
 struct clodConfig
 {
@@ -215,7 +217,7 @@ namespace clod
 struct Cluster
 {
 	size_t vertices;
-	std::vector<unsigned int> indices;
+	clodVector<unsigned int> indices;
 
 	int group;
 	int refined;
@@ -223,7 +225,7 @@ struct Cluster
 	clodBounds bounds;
 };
 
-static clodBounds boundsCompute(const clodMesh& mesh, const std::vector<unsigned int>& indices, float error)
+static clodBounds boundsCompute(const clodMesh& mesh, const clodVector<unsigned int>& indices, float error)
 {
 	meshopt_Bounds bounds = meshopt_computeClusterBounds(&indices[0], indices.size(), mesh.vertex_positions, mesh.vertex_count, mesh.vertex_positions_stride);
 
@@ -254,26 +256,26 @@ static clodBounds boundsMerge(const clodBounds* bounds, size_t count, size_t str
 	return result;
 }
 
-static clodBounds mergeGroups(const std::vector<Cluster>& clusters, const std::vector<int>& group)
+static clodBounds mergeGroups(const clodVector<Cluster>& clusters, const clodVector<int>& group)
 {
-	std::vector<clodBounds> bounds(group.size());
+	clodVector<clodBounds> bounds(group.size());
 	for (size_t j = 0; j < group.size(); ++j)
 		bounds[j] = clusters[group[j]].bounds;
 
 	return boundsMerge(bounds.data(), bounds.size(), sizeof(clodBounds));
 }
 
-static std::vector<Cluster> clusterize(const clodConfig& config, const clodMesh& mesh, const unsigned int* indices, size_t index_count)
+static clodVector<Cluster> clusterize(const clodConfig& config, const clodMesh& mesh, const unsigned int* indices, size_t index_count)
 {
 	size_t max_meshlets = meshopt_buildMeshletsBound(index_count, config.max_vertices, config.min_triangles);
 
-	std::vector<meshopt_Meshlet> meshlets(max_meshlets);
-	std::vector<unsigned int> meshlet_vertices(index_count);
+	clodVector<meshopt_Meshlet> meshlets(max_meshlets);
+	clodVector<unsigned int> meshlet_vertices(index_count);
 
 #if MESHOPTIMIZER_VERSION < 1000
-	std::vector<unsigned char> meshlet_triangles(index_count + max_meshlets * 3); // account for 4b alignment
+	clodVector<unsigned char> meshlet_triangles(index_count + max_meshlets * 3); // account for 4b alignment
 #else
-	std::vector<unsigned char> meshlet_triangles(index_count);
+	clodVector<unsigned char> meshlet_triangles(index_count);
 #endif
 
 	if (config.cluster_spatial)
@@ -285,7 +287,7 @@ static std::vector<Cluster> clusterize(const clodConfig& config, const clodMesh&
 		    mesh.vertex_positions, mesh.vertex_count, mesh.vertex_positions_stride,
 		    config.max_vertices, config.min_triangles, config.max_triangles, 0.f, config.cluster_split_factor));
 
-	std::vector<Cluster> clusters(meshlets.size());
+	clodVector<Cluster> clusters(meshlets.size());
 
 	for (size_t i = 0; i < meshlets.size(); ++i)
 	{
@@ -313,13 +315,13 @@ static std::vector<Cluster> clusterize(const clodConfig& config, const clodMesh&
 	return clusters;
 }
 
-static std::vector<std::vector<int> > partition(const clodConfig& config, const clodMesh& mesh, const std::vector<Cluster>& clusters, const std::vector<int>& pending, const std::vector<unsigned int>& remap)
+static clodVector<clodVector<int> > partition(const clodConfig& config, const clodMesh& mesh, const clodVector<Cluster>& clusters, const clodVector<int>& pending, const clodVector<unsigned int>& remap)
 {
 	if (pending.size() <= config.partition_size)
 		return {pending};
 
-	std::vector<unsigned int> cluster_indices;
-	std::vector<unsigned int> cluster_counts(pending.size());
+	clodVector<unsigned int> cluster_indices;
+	clodVector<unsigned int> cluster_counts(pending.size());
 
 	// copy cluster index data into a flat array for partitioning
 	size_t total_index_count = 0;
@@ -339,21 +341,21 @@ static std::vector<std::vector<int> > partition(const clodConfig& config, const
 	}
 
 	// partition clusters into groups; the output is a partition id per cluster
-	std::vector<unsigned int> cluster_part(pending.size());
+	clodVector<unsigned int> cluster_part(pending.size());
 	size_t partition_count = meshopt_partitionClusters(&cluster_part[0], &cluster_indices[0], cluster_indices.size(), &cluster_counts[0], cluster_counts.size(),
 	    config.partition_spatial ? mesh.vertex_positions : NULL, remap.size(), mesh.vertex_positions_stride, config.partition_size);
 
 	// preallocate partitions for worst case
-	std::vector<std::vector<int> > partitions(partition_count);
+	clodVector<clodVector<int> > partitions(partition_count);
 	for (size_t i = 0; i < partition_count; ++i)
 		partitions[i].reserve(config.partition_size + config.partition_size / 3);
 
-	std::vector<unsigned int> partition_remap;
+	clodVector<unsigned int> partition_remap;
 
 	if (config.partition_sort)
 	{
 		// compute partition points for sorting; any representative point will do, we use last cluster center for simplicity
-		std::vector<float> partition_point(partition_count * 3);
+		clodVector<float> partition_point(partition_count * 3);
 		for (size_t i = 0; i < pending.size(); ++i)
 			memcpy(&partition_point[cluster_part[i] * 3], clusters[pending[i]].bounds.center, sizeof(float) * 3);
 
@@ -369,7 +371,7 @@ static std::vector<std::vector<int> > partition(const clodConfig& config, const
 	return partitions;
 }
 
-static void lockBoundary(std::vector<unsigned char>& locks, const std::vector<std::vector<int> >& groups, const std::vector<Cluster>& clusters, const std::vector<unsigned int>& remap, const unsigned char* vertex_lock)
+static void lockBoundary(clodVector<unsigned char>& locks, const clodVector<clodVector<int> >& groups, const clodVector<Cluster>& clusters, const clodVector<unsigned int>& remap, const unsigned char* vertex_lock)
 {
 	// for each remapped vertex, use bit 7 as temporary storage to indicate that the vertex has been used by a different group previously
 	for (size_t i = 0; i < locks.size(); ++i)
@@ -424,10 +426,10 @@ struct SloppyVertex
 	unsigned int id;
 };
 
-static void simplifyFallback(std::vector<unsigned int>& lod, const clodMesh& mesh, const std::vector<unsigned int>& indices, const std::vector<unsigned char>& locks, size_t target_count, float* error)
+static void simplifyFallback(clodVector<unsigned int>& lod, const clodMesh& mesh, const clodVector<unsigned int>& indices, const clodVector<unsigned char>& locks, size_t target_count, float* error)
 {
-	std::vector<SloppyVertex> subset(indices.size());
-	std::vector<unsigned char> subset_locks(indices.size());
+	clodVector<SloppyVertex> subset(indices.size());
+	clodVector<unsigned char> subset_locks(indices.size());
 
 	lod.resize(indices.size());
 
@@ -458,12 +460,12 @@ static void simplifyFallback(std::vector<unsigned int>& lod, const clodMesh& mes
 		lod[i] = subset[lod[i]].id;
 }
 
-static std::vector<unsigned int> simplify(const clodConfig& config, const clodMesh& mesh, const std::vector<unsigned int>& indices, const std::vector<unsigned char>& locks, size_t target_count, float* error)
+static clodVector<unsigned int> simplify(const clodConfig& config, const clodMesh& mesh, const clodVector<unsigned int>& indices, const clodVector<unsigned char>& locks, size_t target_count, float* error)
 {
 	if (target_count > indices.size())
 		return indices;
 
-	std::vector<unsigned int> lod(indices.size());
+	clodVector<unsigned int> lod(indices.size());
 
 	unsigned int options = meshopt_SimplifySparse | meshopt_SimplifyErrorAbsolute | (config.simplify_permissive ? meshopt_SimplifyPermissive : 0) | (config.simplify_regularize ? meshopt_SimplifyRegularize : 0);
 
@@ -518,9 +520,9 @@ static std::vector<unsigned int> simplify(const clodConfig& config, const clodMe
 	return lod;
 }
 
-static int outputGroup(const clodConfig& config, const clodMesh& mesh, const std::vector<Cluster>& clusters, const std::vector<int>& group, const clodBounds& simplified, int depth, void* output_context, clodOutput output_callback)
+static int outputGroup(const clodConfig& config, const clodMesh& mesh, const clodVector<Cluster>& clusters, const clodVector<int>& group, const clodBounds& simplified, int depth, void* output_context, clodOutput output_callback)
 {
-	std::vector<clodCluster> group_clusters(group.size());
+	clodVector<clodCluster> group_clusters(group.size());
 
 	for (size_t i = 0; i < group.size(); ++i)
 	{
@@ -606,10 +608,10 @@ size_t clodBuild(clodConfig config, clodMesh mesh, void* output_context, clodOut
 	assert(mesh.attribute_count * sizeof(float) <= mesh.vertex_attributes_stride);
 	assert(mesh.attribute_protect_mask < (1u << (mesh.vertex_attributes_stride / sizeof(float))));
 
-	std::vector<unsigned char> locks(mesh.vertex_count);
+	clodVector<unsigned char> locks(mesh.vertex_count);
 
 	// for cluster connectivity, we need a position-only remap that maps vertices with the same position to the same index
-	std::vector<unsigned int> remap(mesh.vertex_count);
+	clodVector<unsigned int> remap(mesh.vertex_count);
 	meshopt_generatePositionRemap(&remap[0], mesh.vertex_positions, mesh.vertex_count, mesh.vertex_positions_stride);
 
 	// set up protect bits on UV seams for permissive mode
@@ -628,13 +630,13 @@ size_t clodBuild(clodConfig config, clodMesh mesh, void* output_context, clodOut
 	}
 
 	// initial clusterization splits the original mesh
-	std::vector<Cluster> clusters = clusterize(config, mesh, mesh.indices, mesh.index_count);
+	clodVector<Cluster> clusters = clusterize(config, mesh, mesh.indices, mesh.index_count);
 
 	// compute initial precise bounds; subsequent bounds will be using group-merged bounds
 	for (Cluster& cluster : clusters)
 		cluster.bounds = boundsCompute(mesh, cluster.indices, 0.f);
 
-	std::vector<int> pending(clusters.size());
+	clodVector<int> pending(clusters.size());
 	for (size_t i = 0; i < clusters.size(); ++i)
 		pending[i] = int(i);
 
@@ -643,7 +645,7 @@ size_t clodBuild(clodConfig config, clodMesh mesh, void* output_context, clodOut
 	// merge and simplify clusters until we can't merge anymore
 	while (pending.size() > 1)
 	{
-		std::vector<std::vector<int> > groups = partition(config, mesh, clusters, pending, remap);
+		clodVector<clodVector<int> > groups = partition(config, mesh, clusters, pending, remap);
 
 		pending.clear();
 
@@ -653,7 +655,7 @@ size_t clodBuild(clodConfig config, clodMesh mesh, void* output_context, clodOut
 		// every group needs to be simplified now
 		for (size_t i = 0; i < groups.size(); ++i)
 		{
-			std::vector<unsigned int> merged;
+			clodVector<unsigned int> merged;
 			merged.reserve(groups[i].size() * config.max_triangles * 3);
 			for (size_t j = 0; j < groups[i].size(); ++j)
 				merged.insert(merged.end(), clusters[groups[i][j]].indices.begin(), clusters[groups[i][j]].indices.end());
@@ -665,7 +667,7 @@ size_t clodBuild(clodConfig config, clodMesh mesh, void* output_context, clodOut
 			clodBounds bounds = mergeGroups(clusters, groups[i]);
 
 			float error = 0.f;
-			std::vector<unsigned int> simplified = simplify(config, mesh, merged, locks, target_size, &error);
+			clodVector<unsigned int> simplified = simplify(config, mesh, merged, locks, target_size, &error);
 			if (simplified.size() > merged.size() * config.simplify_threshold)
 			{
 				bounds.error = FLT_MAX; // terminal group, won't simplify further
@@ -681,9 +683,9 @@ size_t clodBuild(clodConfig config, clodMesh mesh, void* output_context, clodOut
 
 			// discard clusters from the group - they won't be used anymore
 			for (size_t j = 0; j < groups[i].size(); ++j)
-				clusters[groups[i][j]].indices = std::vector<unsigned int>();
+				clusters[groups[i][j]].indices = clodVector<unsigned int>();
 
-			std::vector<Cluster> split = clusterize(config, mesh, simplified.data(), simplified.size());
+			clodVector<Cluster> split = clusterize(config, mesh, simplified.data(), simplified.size());
 
 			for (Cluster& cluster : split)
 			{
@@ -737,8 +739,8 @@ size_t clodBuildHierarchy(clodNode* nodes, const clodGroup* groups, size_t group
 	// reserve space for per-level roots
 	size_t offset = level_count;
 
-	std::vector<clodNode> row(group_count);
-	std::vector<unsigned int> order(group_count);
+	clodVector<clodNode> row(group_count);
+	clodVector<unsigned int> order(group_count);
 
 	for (size_t level = 0; level < level_count; ++level)
 	{
